admin/src/main/java/org/bahmni/module/admin/observation/handler/Form2CSVObsHandler.java

```diff
@@ -82,4 +119,75 @@ private void setFormNamespaceAndFieldPath(List<EncounterTransaction.Observation>
+        setFormNamespaceAndFieldPathForAddmoreObs(form2Observations, form2CSVHeaderParts, addmoreCSVObservations);
+    }
+
+    private List<KeyValue> processMultipleValues(EncounterRow encounterRow, List<EncounterTransaction.Observation> form2Observations, KeyValue form2CSVObservation, List<String> form2CSVHeaderParts, List<String> multipleValues) throws ParseException {
+        List<KeyValue> form2CSVObservations = new ArrayList<>();
+        for (String value : multipleValues) {
+            KeyValue newForm2CSVObservation = new KeyValue();
+            newForm2CSVObservation.setKey(form2CSVObservation.getKey());
+            newForm2CSVObservation.setValue(value.trim());
+            form2CSVObservations.add(newForm2CSVObservation);
+        }
+        csvObservationHelper.createObservations(form2Observations, encounterRow.getEncounterDate(),
+                form2CSVObservations, getConceptNames(form2CSVHeaderParts));
+        return form2CSVObservations;
+    }
+
+    private void setFormNamespaceAndFieldPathForMultiSelectObs(List<EncounterTransaction.Observation> form2Observations, List<String> form2CSVHeaderParts, List<KeyValue> addmoreForm2CSVObservations) {
+        if (isEmpty(form2Observations)) {
+            return;
+        }
+        int prevObsCount = form2Observations.size() - addmoreForm2CSVObservations.size();
+        for(int i = 0; i < addmoreForm2CSVObservations.size(); i++) {
+            final EncounterTransaction.Observation observation = form2Observations.get(prevObsCount + i);
+            final String formFieldPath = formFieldPathService.getFormFieldPath(form2CSVHeaderParts);
+            observation.setFormFieldPath(formFieldPath);
+            observation.setFormNamespace(FORM_NAMESPACE);
+        }
+    }
+
+    private void setFormNamespaceAndFieldPathForAddmoreObs(List<EncounterTransaction.Observation> form2Observations, List<String> form2CSVHeaderParts, List<KeyValue> addmoreForm2CSVObservations) {
+        if (isEmpty(form2Observations)) {
+            return;
+        }
+        int prevObsCount = form2Observations.size() - addmoreForm2CSVObservations.size();
+        final String formFieldPath = formFieldPathService.getFormFieldPath(form2CSVHeaderParts);
+        String[] tokens = formFieldPath.split(FORM_FIELD_PATH_SEPARATOR);
+        int formFieldPathPosition = tokens.length;
+        formFieldPathPosition = tokens.length - 1;
+        String path = tokens[formFieldPathPosition];
+        String controlIdPrefix = path.split("-")[0];
+
+        for(int i = 0; i < addmoreForm2CSVObservations.size(); i++) {
+            final EncounterTransaction.Observation observation = form2Observations.get(prevObsCount + i);
+            tokens[formFieldPathPosition] = controlIdPrefix + "-" + i;
+            observation.setFormFieldPath(String.join(FORM_FIELD_PATH_SEPARATOR, tokens));
+            observation.setFormNamespace(FORM_NAMESPACE);
+        }
+    }
```

Comment:
This code is about parsing Form field locations -- and can be externalised into its own class with its own test cases -- and can be reused/extended independent of Form2CSVObsHandler. I would recommend extracting it for making current class easier to read, and have less logic.

Reply:
Thanks Gurpreet for reviewing. I have refactored this to a new Class.